packages/protoplugin/src/ecmascript/transpile.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright 2021-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import type { FileInfo } from "./generated-file.js";
+import ts from "typescript";
+import {
+  createDefaultMapFromNodeModules,
+  createSystem,
+  createVirtualCompilerHost,
+} from "@typescript/vfs";
+
+/* eslint-disable import/no-named-as-default-member */
+
+function createTranspiler(options: ts.CompilerOptions, files: FileInfo[]) {
+  const fsMap = createDefaultMapFromNodeModules({
+    target: ts.ScriptTarget.ES2015,
+  });
+
+  files.forEach((file) => {
+    fsMap.set(file.name, file.content);
+  });
+
+  const system = createSystem(fsMap);
+  const host = createVirtualCompilerHost(system, options, ts);
+
+  return ts.createProgram({
+    rootNames: [...fsMap.keys()],
+    options,
+    host: host.compilerHost,
+  });
+}
+
+export function transpile(
+  files: FileInfo[],
+  transpileJs: boolean,
+  transpileDts: boolean
+): FileInfo[] {
+  // TODO - Need more options here by default
+  const options: ts.CompilerOptions = {
+    declaration: transpileDts,
+    emitDeclarationOnly: transpileDts && !transpileJs,
+  };
```

Comment:
We do. I think the general goal here should be to emit code, even if it is not completely correct.

Reply:
I added the default compiler options generated by `tsc --init` on 4.5.2.  Let me know if you think there are others we should include.

Reply:
Synced offline - we need to do the opposite here, and be very lenient.

I think this should work:

```ts
    // Type checking
    strict: false,

    // modules
    module: ts.ModuleKind.ES2020,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    noResolve: true,
    resolveJsonModule: false,

    // emit
    emitBOM: false,
    importsNotUsedAsValues: ts.ImportsNotUsedAsValues.Preserve,
    newLine: ts.NewLineKind.LineFeed,
    preserveValueImports: false,

    // JavaScript Support
    allowJs: true,
    checkJs: false,

    // Language and Environment
    lib: [],
    moduleDetection: "force",
    target: ts.ScriptTarget.ES2017,

    // Completeness
    skipLibCheck: true,
    skipDefaultLibCheck: false,
```

I haven't checked if this is the right combination we want for ESM imports. Please adjust as necessary.